File: 2024/04/ts/solver.ts
type Dir = 1 | 0 | -1;

export function solve(input: string): number {
	let count = 0;

	const lines = input.split("\n");
	const rows = lines.length;
	const cols = lines[0].length;

	for (let row = 1; row < rows - 1; row++) {
		for (let col = 1; col < cols - 1; col++) {
			const cell = lines[row][col];
			if (cell !== "A") continue;

			const diagonalOne = lines[row - 1][col - 1] + lines[row + 1][col + 1];
			const diagonalTwo = lines[row - 1][col + 1] + lines[row + 1][col - 1];

			if (diagonalOne.match(/SM|MS/) && diagonalTwo.match(/SM|MS/)) count++;
		}
	}

	return count;
}

File: 2025/04/Sources/04/Solver.swift
import Foundation

public func solve(input: String) -> String {
    let lines = input.split(whereSeparator: \.isNewline).map(String.init)
    var solution = 0

    let rowCount = lines.count

    for (row, line) in lines.enumerated() {
        let colCount = line.count

        for (col, char) in line.enumerated() {
            if char == "." {
                continue
            }
            var adjacentRolls = 0
            for dRow in -1...1 {
                for dCol in -1...1 {
                    if dRow == 0 && dCol == 0 {
                        continue
                    }
                    let newRow = row + dRow
                    let newCol = col + dCol
                    if newRow >= 0 && newRow < rowCount && newCol >= 0 && newCol < colCount {
                        let index = line.index(line.startIndex, offsetBy: newCol)
                        if lines[newRow][index] == "@" {
                            adjacentRolls += 1
                        }
                    }
                }
            }

            if adjacentRolls < 4 {
                solution += 1
            }
        }
    }

    return "Solution: \(solution)"
}
